sw/airborne/arch/chibios/mcu_periph/sys_time_arch.c

```diff
@@ -0,0 +1,145 @@
+ *
+ * Mostly empty functions for Paparazzi compatibility,
+ * since ChibiOS uses different system time functions.
+ */
+#include "mcu_periph/sys_time_arch.h"
+#include BOARD_CONFIG
+#include <ch.h>
+#include "led.h"
+
+/*
+ * Sys_tick handler thread
+ */
+static void thd_sys_tick(void *arg);
+static THD_WORKING_AREA(wa_thd_sys_tick, 1024);
+static void sys_tick_handler(void);
+
+void sys_time_arch_init(void)
+{
+
+  sys_time.cpu_ticks_per_sec = STM32_SYSCLK;
+
+  /* cpu ticks per desired sys_time timer step */
+  sys_time.resolution_cpu_ticks = (uint32_t)(sys_time.resolution * sys_time.cpu_ticks_per_sec + 0.5);
+
+  // Create thread (PRIO should be higher than AP threads
+  chThdCreateStatic(wa_thd_sys_tick, sizeof(wa_thd_sys_tick),
+      NORMALPRIO+2, thd_sys_tick, NULL);
+
+}
+
+/**
+ * Get the time in microseconds since startup.
+ * WARNING: overflows after 70min!
+ * @return microseconds since startup as uint32_t
+ */
+uint32_t get_sys_time_usec(void)
+{
+  return sys_time.nb_sec * 1000000 +
+         usec_of_sys_time_ticks(sys_time.nb_sec_rem) +
+         usec_of_sys_time_ticks(chVTGetSystemTime() - sys_time.nb_tick);
+}
+
+uint32_t get_sys_time_msec(void)
+{
+  return sys_time.nb_sec * 1000 +
+         msec_of_sys_time_ticks(sys_time.nb_sec_rem) +
+         msec_of_sys_time_ticks(chVTGetSystemTime() - sys_time.nb_tick);
+}
+
+/**
+ * sys_time_usleep(uint32_t us)
+ *
+ * using intermediate 64 bits variable to avoid wrapping
+ *
+ * max sleep time is around 10 days (2^32 / CH_CFG_ST_FREQUENCY) at 10kHz
+ */
+void sys_time_usleep(uint32_t us)
+{
+  uint64_t wait_st = ((uint64_t)us * CH_CFG_ST_FREQUENCY) / 1000000UL;
+  chThdSleep((systime_t)wait_st);
+}
+
+void sys_time_msleep(uint16_t ms)
+{
+  uint64_t wait_st = ((uint64_t)ms * CH_CFG_ST_FREQUENCY) / 1000UL;
+  chThdSleep((systime_t)wait_st);
+}
+
+void sys_time_ssleep(uint8_t s)
+{
+  chThdSleep(S2ST(s));
+}
+
+/*
+ * Sys_tick thread
+ */
+static __attribute__((noreturn)) void thd_sys_tick(void *arg)
+{
+  (void) arg;
+  chRegSetThreadName("sys_tick_handler");
+
+  while (TRUE) {
+    sys_tick_handler();
+    chThdSleepMilliseconds(1);
+  }
+}
+
+static void sys_tick_handler(void)
+{
+  /* current time in sys_ticks */
+  sys_time.nb_tick = chVTGetSystemTime();
+  /* max time is 2^32 / CH_CFG_ST_FREQUENCY, i.e. around 10 days at 10kHz */
+  uint32_t sec = sys_time.nb_tick / CH_CFG_ST_FREQUENCY;
+#ifdef SYS_TIME_LED
+  if (sec > sys_time.nb_sec) {
+    LED_TOGGLE(SYS_TIME_LED);
+  }
+#endif
+  sys_time.nb_sec = sec;
+  sys_time.nb_sec_rem = sys_time.nb_tick - sys_time_ticks_of_sec(sys_time.nb_sec);
+
+  /* advance virtual timers */
+  for (unsigned int i = 0; i < SYS_TIME_NB_TIMER; i++) {
+    if (sys_time.timer[i].in_use &&
+        sys_time.nb_tick >= sys_time.timer[i].end_time) {
+      sys_time.timer[i].end_time += sys_time.timer[i].duration;
+      sys_time.timer[i].elapsed = TRUE;
+      /* call registered callbacks, WARNING: they will be executed in the sys_time thread! */
```

Comment:
Sure, but actually not used in current AP code. This was initially done for some specific firmwares and test programs.
So keeping it for now, but would indeed special protection.